site/content/docs/user/configuration.md

```diff
@@ -281,10 +281,71 @@ nodes:
   image: kindest/node:v1.16.4@sha256:b91a2c2317a000f3a783489dfb755064177dbc3a0b2f4147d50f04825d016f55
 {{< /codeFromInline >}}
 
-[Reference](https://kind.sigs.k8s.io/docs/user/quick-start/#creating-a-cluster) 
+[Reference](https://kind.sigs.k8s.io/docs/user/quick-start/#creating-a-cluster)
 
 **Note**: Kubernetes versions are expressed as x.y.z, where x is the major version, y is the minor version, and z is the patch version, following [Semantic Versioning](https://semver.org/) terminology. For more information, see [Kubernetes Release Versioning.](https://github.com/kubernetes/sig-release/blob/master/release-engineering/versioning.md#kubernetes-release-versioning)
 
+### GPU Support
+
+There are two ways to map GPUs in to a KinD cluster. The first is using the `devices` API and the second is using the `extraMounts` API.
+
+#### Using the Devices API
+
+As a pre-requisite you install the [NVIDIA Container Toolkit](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html) installed on the host.
+
+Using `devices` for GPU support requires Docker v25 or later. A [CDI specification](https://github.com/container-orchestrated-devices/container-device-interface) will need to be generated for your device. For Nvidia GPU devices see notes [here.](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/install-guide.html#container-device-interface-cdi-support)
+
+GPU devices can be mapped to Kind node copntainers with the devices API:
+
+All GPUs mapped to a single control-plane:
+
+{{< codeFromInline lang="yaml" >}}
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+nodes:
+- role: control-plane
+  devices:
+  - "nvidia.com/gpu=all"
+{{< /codeFromInline >}}
+
+Specific GPUs mapped to specific worker nodes based on index:
+
+{{< codeFromInline lang="yaml" >}}
+kind: Cluster
+apiVersion: kind.x-k8s.io/v1alpha4
+nodes:
+- role: control-plane
+- role: worker
+  devices:
+  - "nvidia.com/gpu=0"
```

Comment:
Note that this will *not* give you access to just GPU 0, but rather *all* GPUs on the system.

This is because --privileged is passed to the docker call that creates the node, thus giving it access to *all* devices under /dev (including all GPU devices).

Individual GPU injection will only be supported once if/when kind is able to start nodes without --privileged.

Reply:
Ok, this reduces the value of adding an API for CDI devices in my opinion.

Reply:
To be clear, this is the same for all other mechanisms too. The `--privileged` flag will mean that any container will have access to all devices and the only value add is that the driver libraries are also injected.

Reply:
> Ok, this reduces the value of adding an API for CDI devices in my opinion.

This isn't true. The use of the device mounts uses our "legacy" stack which has a number of shortcomings that using CDI addresses. It also means that users can access devices from other vendors such as Intel or their own device / resource definitions.

Reply:
> the only value add is that the driver libraries are also injected.

I don't supposed we can just trigger driver library injection directly in a simple way and reduce the API surface to "yes I'd like the GPU drivers"?

Reply:
Now that CDI support is available in docker, I wanted to revive this thread.

Can someone clarify why `--privileged` is needed in `kind` over something `--cap-add SYS_ADMIN`? As far as I understand things `--privileged` is basically just `(--cap-add SYS_ADMIN` + access to all devices)`. Are there specific devices that `kind` needs access to or is it really just about getting `CAP_SYS_ADMIN` privileges?

Reply:
not my area of expertise, but there are some places you do a lot of "privileged things"

https://github.com/kubernetes-sigs/kind/blob/main/images/base/files/usr/local/bin/entrypoint

https://github.com/kubernetes-sigs/kind/blob/7c3e01fc1c25d833dd9938789a8f26b184472e8f/pkg/cluster/internal/providers/docker/node.go#L105-L108

I'm not sure but we may need NET_ADMIN too

Reply:
It's been a really long time since I've dug into the specifics, but I don't think that's quite it for `--privileged` which is special cased and not directly replicable with other flags IIRC. I definitely attempted to avoid this and IIRC even cap-add=ALL won't work, docker does a few other things like making /sys RW

Reply:
That is how I understood it as well. It is more than just the capabilities that get added and there is no way to replicate with just other CLI options.

Reply:
I chatted with @aojea at Kubecon the other day and he convinced me that it is impractical to remove `--privileged` when starting the docker containers that wrap a worker node. Some end users may be relying on `--privileged`, even if `kind` itself could be made to enumerate the settings necessary to give it "privileged-like" functionality without actually passing `--privileged`.

Therefore, the long-term solution for this will require running some sort of "post-processing" script that removes access to any GPUs that were not selected for injection. Since `kind` is not intended for production, I don't think we need to actually remove `cgroup` access to these GPUs, but rather just make them inaccessible to any workloads that want to make use of them.

For example, something like the following could be used to remove access to all GPUs except GPU 0 on a DGX A100 server
```
cp /proc/driver/nvidia/params root/gpu-params
sed -i 's/^ModifyDeviceFiles: 1$/ModifyDeviceFiles: 0/' root/gpu-params
mount --bind root/gpu-params /proc/driver/nvidia/params

export GPU_EXCLUDE_LIST=1,2,3,4,5,6,7
eval rm -rf /dev/nvidia{${GPU_EXCLUDE_LIST}}
```

Where the first three lines make it so that calls into `nvidia-smi` / `libnvidia-ml.so` do not attempt to recreate nvidia device nodes or reset their permissions if tampered with. And the last two lines remove the device nodes for all GPUs except GPU 0.

This would need to be extended in order to work for MIG devices, but it's a starting point to begin thinking about what we want to do here to support selective GPU injection into different kind nodes.

Reply:
Yeah, this is what I'd been thinking, it's going to be a bit gross and I'm not sure how to generalize it and give a good user experience. It is immediately possible to do this for testing by `docker exec` to your target nodes and running these steps though.

Reply:
I spent today putting this together:
https://github.com/klueska/kind-with-gpus-examples

It should be a good starting point for using GPUs with kind independent of this PR

Reply:
I've also pinged the docker maintainers that helped us push CDI through to see what their thoughts are here:
https://dockercommunity.slack.com/archives/C04MZQZJE94/p1712086101448089

Reply:
Was pointed to this on that thread:
https://github.com/moby/moby/pull/39702#issuecomment-936707664

Reply:
If https://github.com/moby/moby/issues/47663 becomes available we'll have to version detect to use it to avoid breaking all of the users on older docker installs, and I'll be pretty wary of the branching codepath behaving differently vs --privileged.

We have only once required a newer docker version (when we started to require cgroupns=private be available) which caused a lot of consternation even though that change was aimed at mitigating the runc misc issues that caused kind to very flaky/broken ...

Reply:
It would've been nice when kind was being built, but now we have to consider that people undoubtedly have test environments that depend on being in a --privileged node :/